solver-next/internal/pipe/pipe.go

```diff
@@ -0,0 +1,198 @@
+package pipe
+
+import (
+	"context"
+	"sync"
+	"sync/atomic"
+
+	"github.com/pkg/errors"
+)
+
+type channel struct {
+	OnSendCompletion func()
+	mu               sync.Mutex
+	value            atomic.Value
+	lastValue        interface{}
+}
+
+func (c *channel) Send(v interface{}) {
+	c.value.Store(v)
+	if c.OnSendCompletion != nil {
+		c.OnSendCompletion()
+	}
+}
+
+func (c *channel) Receive() (interface{}, bool) {
+	v := c.value.Load()
+	if c.lastValue == v {
+		return nil, false
+	}
+	c.lastValue = v
+	return v, true
+}
+
+type Pipe struct {
+	Sender              Sender
+	Receiver            Receiver
+	OnReceiveCompletion func()
+	OnSendCompletion    func()
+}
+
+type Request struct {
+	Payload  interface{}
+	Canceled bool
+}
+
+type Sender interface {
+	Request() Request
+	Update(v interface{})
+	Finalize(v interface{}, err error)
+	Status() Status
+}
+
+type Receiver interface {
+	Receive() bool
+	Cancel()
+	Status() Status
+	Request() interface{}
+}
+
+type Status struct {
+	Canceled  bool
+	Completed bool
+	Err       error
+	Value     interface{}
+}
+
+func NewWithFunction(f func(context.Context) (interface{}, error)) (*Pipe, func()) {
+	p := New(Request{})
+
+	ctx, cancel := context.WithCancel(context.TODO())
+
+	p.OnReceiveCompletion = func() {
+		if req := p.Sender.Request(); req.Canceled {
+			cancel()
+		}
+	}
+
+	return p, func() {
+		res, err := f(ctx)
+		if err != nil {
+			p.Sender.Finalize(nil, err)
+			return
+		}
+		p.Sender.Finalize(res, nil)
+	}
+}
+
+func New(req Request) *Pipe {
+	cancelCh := &channel{}
+	roundTripCh := &channel{}
+	pw := &sender{
+		req:         req,
+		recvChannel: cancelCh,
+		sendChannel: roundTripCh,
+	}
+	pr := &receiver{
+		req:         req,
+		recvChannel: roundTripCh,
+		sendChannel: cancelCh,
+	}
+
+	p := &Pipe{
+		Sender:   pw,
+		Receiver: pr,
+	}
+
+	cancelCh.OnSendCompletion = func() {
+		v, ok := cancelCh.Receive()
+		if ok {
+			pw.setRequest(v.(Request))
+		}
+		if p.OnReceiveCompletion != nil {
+			p.OnReceiveCompletion()
+		}
+	}
+
+	roundTripCh.OnSendCompletion = func() {
+		if p.OnSendCompletion != nil {
+			p.OnSendCompletion()
+		}
+	}
+
+	return p
+}
+
+type sender struct {
+	status      Status
+	req         Request
+	recvChannel *channel
+	sendChannel *channel
+	mu          sync.Mutex
+}
+
+func (pw *sender) Status() Status {
+	return pw.status
+}
+
+func (pw *sender) Request() Request {
+	pw.mu.Lock()
+	defer pw.mu.Unlock()
+	return pw.req
```

Comment:
Why only Request is locked?

Reply:
Because the internal `setRequest` is the only method that can be called concurrently (it is called to set the cancellation bits). The public API is single threaded.

Reply:
To clarify, the reader/writer methods as a group are single threaded. A reader can be used concurrently with the writer.